### PR TITLE
Grafana installation

### PR DESCRIPTION
**Pull request by mikenairn**
Add observability kustomization

Adds kustomization files to install the observability stack for kcp-glbc (prometheus and grafana). On OpenShift this relies on the pre-configured monitoring stack using user workload monitoring and installs grafana only with an auth proxy for OpenShift login. On k8s clusters, it installs the kube-prometheus stack (prometheus) and grafana with basic login.

Remove arg from deployment (--glbc-kubeconfig=/tmp/glbc/kubeconfig). The default behaviour now expects the kcp glbc to be running on the same cluster as cert-manager etc..

Update local-setup-add-crc-cluster (Enable cluster monitoring)
Update local-setup script so that the glbc cluster has the ingress controller installed.

**Verification**

Terminal 1 (Run KCP)
```
make local-setup
```

Terminal 2 (OpenShift CRC)
```
//Create crc vm with monitoring enabled
$ ./utils/local-setup-add-crc-cluster.sh

//Deploy kcp-glbc
$ kubectl config use-context crc-admin
$ make deploy

//Deploy Observability stack (Grafana)
$ SECRET=`oc get secret -n openshift-user-workload-monitoring | grep  prometheus-user-workload-token | head -n 1 | awk '{print $1 }'`
$ TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o json | jq -r '.data.token') | base64 -d`
$ sed -i "s/CHANGEME/$TOKEN/" config/observability/openshift/grafana/prometheus-grafanadatasource.yaml
$ ./bin/kustomize build config/observability/openshift/ | kubectl apply -f -
//Note: You might need to do this a couple of times

// Deploy Pod Monitor for kcp-glbc
$ ./bin/kustomize build config/prometheus/ | kubectl -n kcp-glbc apply -f -

// Check Prometheus
$ kubectl -n openshift-user-workload-monitoring get secret prometheus-user-workload -ojson | jq -r '.data["prometheus.yaml.gz"]' | base64 -d | gunzip | grep kcp-glbc
- job_name: podMonitor/kcp-glbc/kcp-glbc-controller-manager/0
      - kcp-glbc
    regex: kcp-glbc
    replacement: kcp-glbc/kcp-glbc-controller-manager
    replacement: kcp-glbc

// Check Grafana
$ kubectl get deployments -n kcp-glbc-observability
NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
grafana-deployment                    1/1     1            1           54s
grafana-operator-controller-manager   1/1     1            1           2m14s
$ kubectl get routes -n kcp-glbc-observability
NAME      HOST/PORT                                         PATH   SERVICES          PORT    TERMINATION   WILDCARD
grafana   grafana-kcp-glbc-observability.apps-crc.testing          grafana-service   https   reencrypt     None

Console should be accessible from https://grafana-kcp-glbc-observability.apps-crc.testing (OpenShift login user: kubadmin)
```
Terminal 2 (K8s kind)
```
// Deploy kcp-glbc
$ kubectl config use-context kind-kcp-cluster-glbc-control
$ make deploy

// Deploy Observability stack (Prometheus and Grafana)
$  ./bin/kustomize build config/observability/kubernetes/ | kubectl apply --force-conflicts  --server-side -f -
//Note: You might need to do this a couple of times

// Deploy Pod Monitor for kcp-glbc
./bin/kustomize build config/prometheus/ | kubectl -n kcp-glbc apply -f -

// Check Prometheus
$ kubectl -n monitoring get secret prometheus-k8s -o json | jq -r '.data["prometheus.yaml.gz"]'| base64 -d | gunzip | grep kcp-glbc
- job_name: podMonitor/kcp-glbc/kcp-glbc-controller-manager/0
      - kcp-glbc
    regex: (kcp-glbc);true
    replacement: kcp-glbc/kcp-glbc-controller-manager

$ docker ps --format '{{json .}}' | jq -e 'select(.Names == "kcp-cluster-glbc-control-control-plane").Ports'
"0.0.0.0:8083->80/tcp, 0.0.0.0:8446->443/tcp, 127.0.0.1:34661->6443/tcp"

$ curl -s http://prometheus.127.0.0.1.nip.io:8083/api/v1/targets | jq -e '.data.activeTargets[] | select(.labels.job == "kcp-glbc/kcp-glbc-controller-manager").labels'
{
  "container": "manager",
  "endpoint": "metrics",
  "instance": "10.244.0.11:8080",
  "job": "kcp-glbc/kcp-glbc-controller-manager",
  "namespace": "kcp-glbc",
  "pod": "kcp-glbc-controller-manager-5dcfb4748-hgv9t"
}

Console should be accessible from http://prometheus.127.0.0.1.nip.io:8083

// Check Grafana
$ kubectl get deployments -n kcp-glbc-observability
NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
grafana-deployment                    1/1     1            1           54s
grafana-operator-controller-manager   1/1     1            1           2m14s

Console should be accessible from http://grafana.127.0.0.1.nip.io:8083 (user:admin, password: admin)
```

### CODE REVIEW

**Comment by maleck13** [comment]
close #93

File: utils/local-setup.sh [hunk 1]
```diff
@@ -170,13 +170,16 @@ export KUBECONFIG=.kcp/admin.kubeconfig
 echo "Creating HCG Workspace"
 ${KUBECTL_KCP_BIN} workspace create kcp-glbc --enter
 
+echo "Waiting 15 seconds..."
+sleep 15
+
```
Comment (mikenairn) [inline utils/local-setup.sh:175]: Was seeing a lot of issues with API's not being ready (apiexport, apiresourceschema), adding this crude delay seemed to help.

File: docs/observability/grafana.adoc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+[[prerequisites]]
+== Prerequisites
+
+Prometheus instance, configured to integrate with the KCP GLBC instance deployed on the same cluster, see link:monitoring.adoc[monitoring]
+
+[[kubernetes]]
+=== Kubernetes
+
+Installations on Kubernetes require a prometheus instance deployed and configured on the target cluster. The following will install the
```
Comment (astefanutti) [inline docs/observability/grafana.adoc:15]: nit: one line per sentence.

File: docs/observability/grafana.adoc [hunk 3]
```diff
@@ -0,0 +1,75 @@
+[[grafana]]
+= Grafana
+
+The https://github.com/grafana-operator/grafana-operator[Grafana operator] is used to install and configure grafana as
+part of the KCP GLBC monitoring stack.
+
+[[prerequisites]]
+== Prerequisites
+
+Prometheus instance, configured to integrate with the KCP GLBC instance deployed on the same cluster, see link:monitoring.adoc[monitoring]
```
Comment (astefanutti) [inline docs/observability/grafana.adoc:10]: nit: trailing dot.

File: config/grafana/auth-proxy/grafana.yaml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    enabled: False
+  client:
+    preferService: True
+  serviceAccount:
+    annotations:
+      serviceaccounts.openshift.io/oauth-redirectreference.primary: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"grafana"}}'
+  resources:
+    limits:
+      cpu: 2000m
+      memory: 8000Mi
```
Comment (astefanutti) [inline config/grafana/auth-proxy/grafana.yaml:86]: `Mi` -> `Gi` maybe, and a bit less :)?

File: config/grafana/auth-proxy/grafana.yaml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  ingress:
+    enabled: False
+  client:
+    preferService: True
+  serviceAccount:
+    annotations:
+      serviceaccounts.openshift.io/oauth-redirectreference.primary: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"grafana"}}'
+  resources:
+    limits:
+      cpu: 2000m
```
Comment (astefanutti) [inline config/grafana/auth-proxy/grafana.yaml:85]: `2000m` -> `2`?

File: config/grafana/auth-proxy/service_account.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: grafana-serviceaccount
+  namespace: system
+  annotations:
+    argocd.argoproj.io/compare-options: IgnoreExtraneous
```
Comment (astefanutti) [inline config/grafana/auth-proxy/service_account.yaml:7]: Out of curiosity, why is that needed?

File: config/grafana/grafana.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    name: "grafana-service"
+    labels:
+      app: "grafana"
+      type: "grafana-service"
+  dashboardLabelSelector:
+    - matchExpressions:
+        - { key: app, operator: In, values: [grafana] }
+  resources:
+    limits:
+      cpu: 200m
```
Comment (astefanutti) [inline config/grafana/grafana.yaml:36]: Could that be consistent with the OpenShift version?

**Review by astefanutti — APPROVED** [review]
Very cool!

File: config/grafana/auth-proxy/grafana.yaml (hunk 4 above)
Comment (mikenairn) [reply config/grafana/auth-proxy/grafana.yaml:86]: Will go back through the limits, it's just whatever this example had set https://github.com/grafana-operator/grafana-operator/tree/master/deploy/examples/oauth

File: config/grafana/auth-proxy/service_account.yaml (hunk 6 above)
Comment (mikenairn) [reply config/grafana/auth-proxy/service_account.yaml:7]: probably isn't, came from here https://github.com/grafana-operator/grafana-operator/blob/master/deploy/examples/oauth/service_account.yaml#L7. Will just remove it.

File: config/grafana/grafana.yaml (hunk 7 above)
Comment (mikenairn) [reply config/grafana/grafana.yaml:36]: probably, will update

File: config/grafana/auth-proxy/grafana.yaml (hunk 5 above)
Comment (mikenairn) [reply config/grafana/auth-proxy/grafana.yaml]: Updated to match OS Deployed grafana
